ath79: add support for Phicomm K2T

hackpascal opened the pull request:
This patch adds dts for qca956x and also support for Phicomm K2T

The qca965x.dtsi adds nearly all the necessary components.
Both ath9k AHB and PCIe worked well.

The Phicomm K2T uses MTD partition 'config' to store the mac addresses in JSON format. To extract these fields correctly, a script is introduced:
    /lib/functions/k2t.sh
This script provides a helper function to extract mac addresses, and is used in three places.

Hardware spec of Phicomm K2T:
CPU: QCA9563
DRAM: 64MB DDR2
Flash: 16MB SPI-NOR
Switch: QCA8337
WiFi 5.8GHz: QCA9886

Flash instruction:
Upgrade using serial console + ethernet:

tftp 0x80000000 sysupgrade.bin && erase 0x9f090000 +$filesize && cp.b $fileaddr 0x9f090000 $filesize

Signed-off-by: Weijie Gao <hackpascal@gmail.com>

juppin commented on target/linux/ath79/dts/qca956x.dtsi:
```diff
@@ -0,0 +1,277 @@
+&eth0 {
+	compatible = "qca,qca9560-eth", "syscon";
+
+	pll-data = <0x03000101 0x80000101 0x80001313>;
```
If you want to set pll-data you need to call ath79_set_pll(ag) in [ag71xx_main.c#L637](https://github.com/openwrt/openwrt/blob/master/target/linux/ath79/files/drivers/net/ethernet/atheros/ag71xx/ag71xx_main.c#L632) for qca9560 and also provide pll-reg and pll-handle.
Don´t know if you need it for your ar8337 switch.

Default values for qca9560 will be:
```
arch/mips/ath79/dev-eth.c
...
#define QCA956X_PLL_VAL_1000	0x03000000
#define QCA956X_PLL_VAL_100	0x00000101
#define QCA956X_PLL_VAL_10	0x00001919
...
```

juppin replied:
Don´t think it should be ath79_set_pllval instead of ath79_set_pll... Are you sure?
On ar9344 it works very good with ath79_set_pll!
You should consider to set the default qca956x values in the qca956x.dtsi and overwrite the default in your dts.

981213 commented:
@ptpt52 ath19kcal_patch_mac_crc also fixed checksum in calibration data and it is tested to be working.
As the factory mac address is stored using standard JSON format, it would be better to read it as a JSON. Your using of those commands is more like a 'dirty hack' and may be broken somehow.

ptpt52 commented:
@981213 
mac patch not really works on K2T A1, my test.
I am sure that, mac patch without crc-fix also works on K2T A3

so for generally support both K2T A1/A3
we just fix mac in /etc/hotplug.d/ieee80211/10_fix_wifi_mac script
and no need to patch the mac

981213 commented:
I doubt whether it's OK to use jshn in board config. The generation of /etc/board.json is already using jshn and by loading an extra json data it seemed that we could add some garbage data into board.json (Just an assumption without any test.)

ptpt52 commented:
BTW, I think a simple hack can work very effectively and correctly
introducing an new script for all the ath79 base devices is unnecessary I think.

981213 commented:
@ptpt52 @hackpascal is using an K2T A1 and it's working correctly.

ptpt52 commented:
@981213 ok, I will use this patch script to do a test.

981213 commented:
It's not necessary to introduce a separate script for just one device however that function could be rewritten and put in /lib/functions.sh for a generic json loading from mtd and we might use it for future devices.

ptpt52 commented:
@981213 my test is ok. this mac patch crc works on K2T A1/A3
so the fixup in /etc/hotplug.d/ieee80211/10_fix_wifi_mac is no need

2.4G fixup can be fixup the same way.
using the ath9k_patch_firmware_mac
in /etc/hotplug.d/firmware/10-ath9k-eeprom

all test ok on my pull 
https://github.com/openwrt/openwrt/pull/1031

981213 commented:
@ptpt52 The device support here told ath9k to read eeprom directly from mtd devices and the wifi mac fix for wmac is necessary.

hackpascal committed:
ath79: add support for Phicomm K2T

This patch adds dts for qca956x and also support for Phicomm K2T

The qca965x.dtsi adds nearly all the necessary components.
Both ath9k AHB and PCIe worked well.

The Phicomm K2T uses MTD partition 'config' to store the mac addresses in
JSON format. To extract these fields correctly, a script is introduced:
    /lib/functions/k2t.sh
This script provides a helper function to extract mac addresses, and is used
in three places.

Hardware spec of Phicomm K2T:
CPU: QCA9563
DRAM: 64MB DDR2
Flash: 16MB SPI-NOR
Switch: QCA8337
WiFi 5.8GHz: QCA9886

Flash instruction:
Apply sysupgrade.bin via serial console:

tftp 0x80000000 sysupgrade.bin && erase 0x9f090000 +$filesize && cp.b $fileaddr 0x9f090000 $filesize

Signed-off-by: Weijie Gao <hackpascal@gmail.com>

hackpascal commented:
@981213 I've fixed the possible json issue by using a json namespace for K2T config data.

hackpascal committed:
ath79: ag71xx: fix pll-data setting for ar7242/ar934x/qca955x/qca956x

ar71xx/ar913x series use the old pll registers and settings.

However started from ar7242, a new pll register is introduced and the
pll setting is much simpler.

This can be observed from dev-eth.c from the ar71xx target.

Signed-off-by: Weijie Gao <hackpascal@gmail.com>

lucize commented:
@hackpascal many thanks for qca956x.dtsi !

lucize commented:
@hackpascal did you do a pppoe torrent test ? the 1043v4 is crashing in a few seconds when I try
I had this before on ar71xx but [this](https://github.com/openwrt/openwrt/commit/d40a358136fdc19e6af13921867ed93444c08827) solved the problem

hackpascal commented:
@lucize sorry I don't have an environment to test it.

lucize commented:
@hackpascal I think is ok just to make a local pppoe server (maybe using a virtual machine) and do a iperf3 between a client and the pppoe server

blogic commented:
merged, Thanks !

reee commented:
Thanks for your work! Could you guys merge the phicomm K3 support to openwrt mainline too?

securecrt commented:
I cant see the K2T build in daily snapshot https://downloads.openwrt.org/snapshots/targets/  there is no ath79 directory.

juppin commented:
ath79 is marked as source-only and therefore no builds happen until this is removed
